cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -202,4 +202,51 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
     }
   }
 
+  public void restartAllServiceInstances(HttpRequest request, HttpResponder responder, String serviceName) {
+    restartInstances(responder, serviceName, -1);
+  }
+
+  public void restartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName,
+                                     int instanceId) {
+    restartInstances(responder, serviceName, instanceId);
+  }
+
+  private void restartInstances(HttpResponder responder, String serviceName, int instanceId) {
+    long startTimeMs = System.currentTimeMillis();
+    boolean isSuccess = true;
+    try {
+      if (!serviceManagementMap.containsKey(serviceName)) {
+        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+        return;
+      }
+
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      isSuccess = (instanceId < 0) ? masterServiceManager.restartAllInstances() :
+        masterServiceManager.restartInstance(instanceId);
+
+      if (isSuccess) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        // throw exception to have unified error handling.
+        throw new RuntimeException();
```

Comment:
`masterServiceManager.restartAllInstances()` already logs, so I don't think it's necessary here. This does need to change though, probably in another PR.